apps/asap-server/test/handlers/research-output/index-handler.test.ts

```diff
@@ -20,36 +22,379 @@ describe('Research Output index handler', () => {
+    expect(algoliaIndexMock.saveObject).toHaveBeenCalledWith({
+      ...researchOutputResponse,
+      objectID: researchOutputResponse.id,
+    });
+  });
+
+  test('Should remove the record Algolia when research-output is unpublished', async () => {
+    const event = unpublishedEvent('ro-1234');
+    await indexHandler(event);
+
+    expect(algoliaIndexMock.deleteObject).toHaveBeenCalledWith(
+      event.detail.payload.id,
+    );
+  });
+
+  test('Should remove the record Algolia when research-output is deleted', async () => {
+    const event = deleteEvent('ro-1234');
+    await indexHandler(event);
+
+    expect(algoliaIndexMock.deleteObject).toHaveBeenCalledWith(
+      event.detail.payload.id,
+    );
+  });
+
+  describe('Should process all events, and not relay on the order', () => {
+    it('receives on correct order the events created and updated', async () => {
+      const roID = 'ro-1234';
+      const createEv = createEvent(roID);
+      const updateEv = updateEvent(roID);
+      const researchOutputResponse = {
+        ...getResearchOutputResponse(),
+        id: roID,
+      };
+
+      researchOutputControllerMock.fetchById.mockResolvedValueOnce({
+        ...researchOutputResponse,
+        title: 'Original title',
+      });
+
+      await indexHandler(createEv);
+      expect(algoliaIndexMock.deleteObject).not.toHaveBeenCalled();
+      expect(algoliaIndexMock.saveObject).toHaveBeenCalledWith({
+        ...researchOutputResponse,
+        title: 'Original title',
+        objectID: researchOutputResponse.id,
+      });
+
+      researchOutputControllerMock.fetchById.mockResolvedValueOnce({
+        ...researchOutputResponse,
+        title: 'New title',
+      });
+
+      await indexHandler(updateEv);
+      expect(algoliaIndexMock.deleteObject).not.toHaveBeenCalled();
+      expect(algoliaIndexMock.saveObject).toHaveBeenCalledTimes(2);
+      expect(algoliaIndexMock.saveObject).toHaveBeenLastCalledWith({
+        ...researchOutputResponse,
+        title: 'New title',
+        objectID: researchOutputResponse.id,
+      });
+    });
+
+    it('receives on reverse order the events created and updated', async () => {
+      const roID = 'ro-1234';
+      const createEv = createEvent(roID);
+      const updateEv = updateEvent(roID);
+      const researchOutputResponse = {
+        ...getResearchOutputResponse(),
+        id: roID,
+      };
+
+      researchOutputControllerMock.fetchById.mockResolvedValueOnce({
+        ...researchOutputResponse,
+        title: 'Updated title',
+      });
+
+      await indexHandler(updateEv);
+      expect(algoliaIndexMock.deleteObject).not.toHaveBeenCalled();
+      expect(algoliaIndexMock.saveObject).toHaveBeenLastCalledWith({
+        ...researchOutputResponse,
+        title: 'Updated title',
+        objectID: researchOutputResponse.id,
+      });
+
+      researchOutputControllerMock.fetchById.mockResolvedValueOnce({
+        ...researchOutputResponse,
+        title: 'Old title',
```

Comment:
this is a bit misleading - how would the second call to the `fetchById` return some 'old title'? I mean it could return anything but you would probably want to say that the `Updated title` is return the second time. In fact I would think that both the calls return the same `Updated title` if we are testing race conditions (so it would be whatever the last state of the resource was).